modules/configuration-annotation-processor/src/test/java/org/apache/ignite/configuration/internal/validation/ValidationUtilTest.java

```diff
@@ -0,0 +1,183 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.configuration.internal.validation;
+
+import java.lang.annotation.Annotation;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.ignite.configuration.annotation.Config;
+import org.apache.ignite.configuration.annotation.ConfigValue;
+import org.apache.ignite.configuration.annotation.ConfigurationRoot;
+import org.apache.ignite.configuration.annotation.NamedConfigValue;
+import org.apache.ignite.configuration.annotation.Value;
+import org.apache.ignite.configuration.internal.RootsNode;
+import org.apache.ignite.configuration.internal.util.ConfigurationUtil;
+import org.apache.ignite.configuration.sample.storage.TestConfigurationStorage;
+import org.apache.ignite.configuration.tree.NamedListView;
+import org.apache.ignite.configuration.validation.ValidationContext;
+import org.apache.ignite.configuration.validation.ValidationIssue;
+import org.apache.ignite.configuration.validation.Validator;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static java.lang.annotation.ElementType.FIELD;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.emptySet;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+/** */
+public class ValidationUtilTest {
+    /** */
+    @Target(FIELD)
+    @Retention(RUNTIME)
+    @interface LeafValidation {
+    }
+
+    /** */
+    @Target(FIELD)
+    @Retention(RUNTIME)
+    @interface InnerValidation {
+    }
+
+    /** */
+    @Target(FIELD)
+    @Retention(RUNTIME)
+    @interface NamedListValidation {
+    }
+
+    /** */
+    @ConfigurationRoot(rootName = "root", storage = TestConfigurationStorage.class)
+    public static class ValidatedRootConfigurationSchema {
+        /** */
+        @InnerValidation
+        @ConfigValue
+        public ValidatedChildConfigurationSchema child;
+
+        /** */
+        @NamedListValidation
+        @NamedConfigValue
+        public ValidatedChildConfigurationSchema elements;
+    }
+
+    /** */
+    @Config
+    public static class ValidatedChildConfigurationSchema {
+        /** */
+        @LeafValidation
+        @Value(hasDefault = true)
+        public String str = "foo";
+    }
+
+    /** */
+    private ValidatedRootNode root;
+
+    /** */
+    @BeforeEach
+    public void before() {
+        root = new ValidatedRootNode();
+
+        ConfigurationUtil.addDefaults(root, root);
+    }
+
+    /** */
+    @AfterEach
+    public void after() {
```

Comment:
Bad habit I think, I'll remove it.
I think I did it automatically because in Ignite 2.x memory would leak, finished test objects are cached somewhere for some reason

Reply:
BTW, what are the rules of "@Before / @BeforeEach"? Should I really split initialization and "preparation" of the field?
It's more about conventions, right?

Reply:
> Should I really split initialization and "preparation" of the field?It's more about conventions, right?

Yes, splitting logic between constructor and `beforeEach` is a matter of taste. I suggested moving the initialization out of here, because then you can make the `root` field `final`.

Reply:
I already moved it. I'll make it final if you wish

Reply:
that's not my wish, I just explained the logic behind my original suggestion =) So this is not at all mandatory